### PR TITLE
feat(Core/Creature): Allow 3 ids per spawn point. TESTING

### PR DESCRIPTION
**Pull request by malcrom**
## Changes Proposed:
-  After a lot of sniffing I was sure that there were no more than 2 entries per spawn point but yesterday found out I was wrong. After discussion we decided to remove chance and replace with a third id to allow 1 of 3 entries to spawn. I will be linking a PR for the devilsaurs in Un'Goro to this PR for testing.

Also if you know a better way to do the random select in creature.cpp please post it here. 

## Issues Addressed:
- Closes 

## SOURCE:
sniffing

## Tests Performed:
- Builds.
- Seems to work so far.

## How to Test the Changes:
1. Test with PR #10171

### CODE REVIEW

**Review by acidmanifesto — APPROVED** [review]

**Comment by pklloveyou** [comment]
Tested

File: src/server/game/Entities/Creature/Creature.cpp [hunk 1]
```diff
@@ -1913,7 +1929,26 @@ void Creature::Respawn(bool force)
...
-                uint32 entry = (roll_chance_f(data->chance_id1)) ? data->id : data->id2;
+                // Add to world
+                uint32 entry = data->id;
+                uint8 ids = 1;
+                if (data->id3) ++ids;
+
+                if (ids)
+                {
+                    uint8 idNumber = urand(0, ids);
+                    switch (idNumber) {
```
Comment (Kitzunu) [inline src/server/game/Entities/Creature/Creature.cpp:1940]: Make this a function so ypu dont have to use the exact same code twice

File: src/server/game/Entities/Creature/CreatureData.h [hunk 2]
```diff
@@ -361,7 +361,7 @@ struct CreatureData
     CreatureData()  { }
     uint32 id{0};                                              // entry in creature_template
```
Comment (Kitzunu) [inline src/server/game/Entities/Creature/CreatureData.h:362]: rename to id1 for concistency

File: src/server/game/Events/GameEventMgr.cpp [hunk 3]
```diff
@@ -533,6 +534,18 @@ void GameEventMgr::LoadFromDB()
...
+                        LOG_ERROR("sql.sql", "Table `game_event_model_equip` have creature (Guid: %u) with equipment_id %u not found in table `creature_equip_template`, set to no equipment.",
+                            guid, newModelEquipSet.equipment_id);
+                        continue;
+                    }
+                    if (entry3 && !sObjectMgr->GetEquipmentInfo(entry3, equipId))
+                    {
+                        LOG_ERROR("sql.sql", "Table `game_event_model_equip` have creature (Guid: %u) with equipment_id %u not found in table `creature_equip_template`, set to no equipment.",
+                            guid, newModelEquipSet.equipment_id);
+                        continue;
+                    }
```
Comment (Kitzunu) [inline src/server/game/Events/GameEventMgr.cpp:548]: I think you can just combine all of these into the first if statement. 

so something like
```cpp
if (!sObjectMgr->GetEquipmentInfo(entry, equipId) || (entry2 && !sObjectMgr->GetEquipmentInfo(entry2, equipId) || (entry3 && !sObjectMgr->GetEquipmentInfo(entry3, equipId))
...
```

And then also refer to the creature id in the error log as we have multiple ids per guid :)

File: src/server/game/Entities/Creature/Creature.cpp (hunk 1 above)
Comment (Kitzunu) [reply src/server/game/Entities/Creature/Creature.cpp:1940]: Creature::GetRandomId() or something

Comment (Kitzunu) [reply src/server/game/Entities/Creature/Creature.cpp:1940]: Just a mock up I thought of
```cpp
entry = GetRandomId(id2, id3);
```
```cpp
Creature::GetRandomId(bool hasId2, bool hasId3)
{
    if (!hasId2)
        return id = data->id1;

    unit8 ids = 0;
    if (hasId2) ++ids;
    if (hasId3) ++ids;

    if (ids)
    {
        uint8 idNumber = urand(0, ids);
        switch (idNumber)
            case 0:
                id = data->id1;
                break;
            case 1:
                id = data->id2;
                break;
            case 2:
                id = data->id3;
                break;
    }

    return id;
}
```

File: src/server/game/Entities/Creature/CreatureData.h (hunk 2 above)
Comment (malcrom) [reply src/server/game/Entities/Creature/CreatureData.h:362]: That's a lot of renaming there.

File: src/server/game/Events/GameEventMgr.cpp (hunk 3 above)
Comment (malcrom) [reply src/server/game/Events/GameEventMgr.cpp:548]: Updated. Will push after compile.

File: src/server/game/Entities/Creature/Creature.cpp (hunk 1 above)
Comment (malcrom) [reply src/server/game/Entities/Creature/Creature.cpp:1940]: Updated. Will push after compile.

File: src/server/game/Entities/Creature/CreatureData.h (hunk 2 above)
Comment (malcrom) [reply src/server/game/Entities/Creature/CreatureData.h]: Done.

File: src/server/game/Entities/Creature/Creature.cpp [hunk 4]
```diff
@@ -3458,3 +3456,33 @@ bool Creature::CanPeriodicallyCallForAssistance() const
 
     return true;
 }
+
+uint32 Creature::GetRandomId(uint32 Id1, uint32 Id2, uint32 Id3)
```
Comment (Kitzunu) [inline src/server/game/Entities/Creature/Creature.cpp:3460]:
```suggestion
uint32 Creature::GetRandomId(uint32 id1, uint32 id2, uint32 id3)
```

File: src/server/database/Database/Implementation/WorldDatabase.cpp [hunk 5]
```diff
@@ -80,10 +80,10 @@ void WorldDatabaseConnection::DoPrepareStatements()
...
     PrepareStatement(WORLD_SEL_WAYPOINT_SCRIPT_BY_ID, "SELECT guid, delay, command, datalong, datalong2, dataint, x, y, z, o FROM waypoint_scripts WHERE id = ?", CONNECTION_SYNCH);
     PrepareStatement(WORLD_SEL_ITEM_TEMPLATE_BY_NAME, "SELECT entry FROM item_template WHERE name = ?", CONNECTION_SYNCH);
-    PrepareStatement(WORLD_SEL_CREATURE_BY_ID, "SELECT guid FROM creature WHERE creature_id1 = ?", CONNECTION_SYNCH);
+    PrepareStatement(WORLD_SEL_CREATURE_BY_ID, "SELECT guid FROM creature WHERE id1 = ?", CONNECTION_SYNCH);
```
Comment (Kitzunu) [inline src/server/database/Database/Implementation/WorldDatabase.cpp:83]:
```suggestion
    PrepareStatement(WORLD_SEL_CREATURE_BY_ID, "SELECT guid FROM creature WHERE id1 = ? OR id2 = ? OR id3 = ?", CONNECTION_SYNCH);
```
This is probably better

**Comment by pklloveyou** [comment]
Tested

File: src/server/game/Globals/ObjectMgr.cpp [hunk 6]
```diff
@@ -2016,36 +2016,37 @@ void ObjectMgr::LoadCreatures()
...
-            LOG_ERROR("sql.sql", "Table `creature` has creature (SpawnId: %u) with non existing creature entry %u in creature_id2 field, skipped.", spawnId, id2);
+            LOG_ERROR("sql.sql", "Table `creature` has creature (SpawnId: %u) with non existing creature entry %u in id2 field, skipped.", spawnId, id2);
             continue;
         }
-        if (!chance || chance > 100)
+        CreatureTemplate const* cInfo3 = GetCreatureTemplate(id3);
+        if (!cInfo3 && id3)
         {
-            LOG_ERROR("sql.sql", "Table `creature` chance_id1 (Value: %u) must be greater than 0 and less than or equal to 100, skipped.", chance);
+            LOG_ERROR("sql.sql", "Table `creature` has creature (SpawnId: %u) with non existing creature entry %u in id3 field, skipped.", spawnId, id3);
```
[the remark this answers is not in the document]
Comment (Kitzunu) [reply src/server/game/Globals/ObjectMgr.cpp:2038]: Yes, they should be skipped if you put a non-valid creature id

[the remark this answers is not in the document]
Comment (Bogir) [reply src/server/game/Globals/ObjectMgr.cpp:2038]: Oh, I noticed removed field chance_id1
ok

**Pull request merged**